conformance/tests/httproute.go

```diff
@@ -0,0 +1,254 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tests
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+	"sigs.k8s.io/gateway-api/conformance/utils/http"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/roundtripper"
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+)
+
+var ConformanceTests []suite.ConformanceTest = []suite.ConformanceTest{{
+	ShortName:   "HTTPRouteSimpleSameNamespace",
+	Description: "A single HTTPRoute in the gateway-conformance-infra namespace",
+	Manifests:   []string{"httproute/simple-same-namespace.yaml"},
+	Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+		ns := v1alpha2.Namespace("gateway-conformance-infra")
+		routeNN := types.NamespacedName{Name: "gateway-conformance-infra-test", Namespace: string(ns)}
+		gwNN := types.NamespacedName{Name: "gateway-same-namespace", Namespace: string(ns)}
+		gwAddr := kubernetes.GatewayAndHTTPRoutesMustBeReady(t, suite.Client, suite.ControllerName, gwNN, routeNN)
+
+		t.Run("Simple HTTP request should reach infra-backend", func(t *testing.T) {
+			t.Logf("Making request to http://%s", gwAddr)
+			httpProto := v1alpha2.HTTPProtocolType
+			cReq, cRes, err := suite.RoundTripper.CaptureRoundTrip(roundtripper.Request{
+				URL:      url.URL{Scheme: "http", Host: gwAddr},
+				Protocol: &httpProto,
+			})
+
+			require.NoErrorf(t, err, "error making request")
+
+			http.ExpectResponse(t, cReq, cRes, http.ExpectedResponse{
+				Request: http.ExpectedRequest{
+					Method: "GET",
+					Path:   "/",
+				},
+				StatusCode: 200,
+				Backend:    "infra-backend",
+				Namespace:  "gateway-conformance-infra",
+			})
+		})
+	},
+}, {
+	ShortName:   "HTTPRouteMatching",
+	Description: "A single HTTPRoute with path matching for different backends",
+	Manifests:   []string{"httproute/matching.yaml"},
+	Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+		ns := "gateway-conformance-infra"
+		routeNN := types.NamespacedName{Name: "httproute-matching", Namespace: ns}
+		gwNN := types.NamespacedName{Name: "gateway-same-namespace", Namespace: ns}
+		gwAddr := kubernetes.GatewayAndHTTPRoutesMustBeReady(t, suite.Client, suite.ControllerName, gwNN, routeNN)
+
+		testCases := []http.ExpectedResponse{{
+			Request:   http.ExpectedRequest{Path: "/"},
+			Backend:   "infra-backend-v1",
+			Namespace: ns,
+		}, {
+			Request:   http.ExpectedRequest{Path: "/example"},
+			Backend:   "infra-backend-v1",
+			Namespace: ns,
+		}, {
+			Request:   http.ExpectedRequest{Path: "/", Headers: map[string]string{"version": "one"}},
```

Comment:
Header usage doesn't work I think. By the time we get here go will have canonicalized the header name. IE should be checking `Version` (or some other fix)

Reply:
Good catch, updated to check "Version" for now, but will work on a better fix that makes the header check case insensitive.